source/extensions/filters/http/cache/hazelcast_http_cache/hazelcast_http_cache.cc

```diff
@@ -0,0 +1,178 @@
+
+HazelcastHttpCache::HazelcastHttpCache(HazelcastHttpCacheConfig config)
+    : unified_(config.unified()),
+      body_partition_size_(ConfigUtil::validPartitionSize(config.body_partition_size())),
+      max_body_size_(ConfigUtil::validMaxBodySize(config.max_body_size(), config.unified())),
+      cache_config_(config) {}
+
+void HazelcastHttpCache::onMissingBody(uint64_t key, int32_t version, uint64_t body_size) {
+  try {
+    if (!tryLock(key)) {
+      // Let lock owner context to recover it.
+      return;
+    }
+    auto header = getHeader(key);
+    if (header && header->version() != version) {
+      // The missed body does not belong to the looked up header. Probably eviction and then
+      // insertion for the header has happened in the meantime. Since new insertion will
+      // override the existing bodies, ignore the cleanup and let orphan bodies (belong to
+      // evicted header, not overridden) be evicted by TTL as well.
+      unlock(key);
+      return;
+    }
+    int body_count = body_size / body_partition_size_;
+    while (body_count >= 0) {
+      accessor_->removeBodyAsync(orderedMapKey(key, body_count--));
+    }
+    accessor_->removeHeader(mapKey(key));
+    unlock(key);
+  } catch (HazelcastClientOfflineException& e) {
+    // see DividedInsertContext::insertHeader for left over locks on a connection failure.
+    ENVOY_LOG(warn, "Hazelcast Connection is offline!");
+  } catch (std::exception& e) {
+    ENVOY_LOG(warn, "Clean up for missing body has failed: {}", e.what());
+  }
+}
+
+void HazelcastHttpCache::onVersionMismatch(uint64_t key, int32_t version, uint64_t body_size) {
+  onMissingBody(key, version, body_size);
+}
+
+void HazelcastHttpCache::start() {
+  if (accessor_ && accessor_->isRunning()) {
+    ENVOY_LOG(warn, "Client is already connected. Cluster name: {}", accessor_->clusterName());
+    return;
+  }
+
+  ClientConfig client_config = ConfigUtil::getClientConfig(cache_config_);
+  client_config.getSerializationConfig().addDataSerializableFactory(
+      HazelcastCacheEntrySerializableFactory::FACTORY_ID,
+      boost::shared_ptr<DataSerializableFactory>(new HazelcastCacheEntrySerializableFactory()));
+
+  if (!accessor_) {
+    accessor_ = std::make_unique<HazelcastClusterAccessor>(
+        *this, std::move(client_config), cache_config_.app_prefix(), body_partition_size_);
+    ENVOY_LOG(debug, "New HazelcastClusterAccessor created.");
+  }
+
+  try {
+    accessor_->connect();
+  } catch (...) {
+    accessor_.reset();
+    throw EnvoyException("Hazelcast Client could not connect to any cluster.");
+  }
+  ENVOY_LOG(info, "HazelcastHttpCache has been started with profile: {}. Max body size: {}.",
+            unified_ ? "UNIFIED"
+                     : "DIVIDED, partition size: " + std::to_string(body_partition_size_),
+            max_body_size_);
+  HazelcastClusterAccessor& cluster_accessor = static_cast<HazelcastClusterAccessor&>(*accessor_);
```

Comment:
How do we know that this is a well-defined cast?

Reply:
The only StorageAccessor implementation in source is HazelcastClusterAccessor and the factory creates the cache accordingly. The reason for having StorageAccessor abstraction is to make it able to test the cache behavior with local storage rather than remote connections. Otherwise at least one Hazelcast instance must be up and running in the CI environment.

Reply:
I think the code would be simpler and its correctness would be more obvious if we moved responseMapName and headerMapName into StorageAccessor, and thus obviated the need for static_cast. That would also remove an impediment to unit-testing HazelcastHttpCache::start with a fake StorageAccessor.

Reply:
Made HazelcastHttpCache::start non virtual and passed accessor as parameter. Factory and tests now use the proper accessor when starting the cache.